client-js/app/scripts/obligation.ts

```diff
@@ -0,0 +1,48 @@
+/**
+ * Obligation is the input side complement of Promise; if one party is promised
+ * a value, then of necessity another party must be obliged to provide the
+ * value or explain why they failed to do so. Obligation has two methods
+ * `resolve` and `reject` which can be used to send a value or an error into a
+ * chain of asynchronous callbacks. Promise is the output side of the chain,
+ * providing access to the asynchronous values by adding callback functions.
+ *
+ * In the traditional ES6 Promise API, resolve and reject functions are only
+ * available as arguments passed in to the executor function we pass to the
+ * Promise constructor. But in many situations we need to store the callbacks
+ * for later use or access them outside the constructor, which Obligation makes
+ * possible. This is especially useful in situations where we need to interface
+ * traditional callback-based code with Promise-based code.
+ *
+ * Compare the scala.concurrent library, which properly separates the input and
+ * output sides of an asynchronous callback chain into two classes:
+ * scala.concurrent.Promise is the input side (equivalent to Obligation here)
+ * and scala.concurrent.Future is the output side (equivalent to Promise here).
+ */
+export class Obligation<R> {
+
+  public promise: Promise<R>;
```

Comment:
Why public? It seems to me the only reason to make this public is so that an asynchronous API can do something like this:
```javascript
somethingAsync(): Promise<A>() {
  var obligation = new Obligation<A>();
  myObligations.push(obligation);
  return obligation.promise;
}
```
If the only point of making the `promise` field public is to return it to the piece of code to which we're making the promise, I'd rather just use a Promise/Obligation factory and keep the `promise` field of `Obligation` private.